### DISCUSSION TITLE
How to find out which elements the node (dof) is on

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#27626] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

hi,

I need to do different things to solutions of different elements, but in continuous finite element method, different elements will share some degrees of freedom (dof), which will result in some solutions being manipulated multiple times. I need to know which elements each Node or dof belongs to, and then do special handling as appropriate. So how can I get these messages when the pointer to the Node or dof is known?

Thank you very much for your attention and I would quite appreciate it if you have any suggestion!

**Comment by GiudGiud** [comment]
Hello

A loop on nodes instead of elements will prevent this double modification, at least for lagrange variables. 
Nodes inherit from the DofInterface, you can get the dof of each variable using the dof_number(variable, system) routine.

Alternatively, you can work with the DoFMap directly

Guillaume

**Reply by hityyds** [reply]
Thank you for your reply. Now I do want to handle it that way, but I need to know which element each Node is on before I can do the corresponding calculation operation. The coefficients in the calculation depend on which element it is on.

**Reply by GiudGiud** [reply, marked as answer]
there is a node_to_elem_map you can get from the mesh to help with that

**Reply by hityyds** [reply]
This seems to be exactly what I need, thank you so much for your help!

**Reply by lindsayad** [reply]
@hityyds please mark this as answered if this answered your question. I will note that we do not assign a node an element owner. As you know several elements can share a given node; we do not pick any of them as the node "owner"

**Reply by hityyds** [reply]
@lindsayad I apologize, I thought I had already marked it as answered. Thank you for the reminder. Regarding whether a node has an element owner, I'm curious how moose assigns material data to nodes at the interfaces of different materials.

**Reply by lindsayad** [reply]
The traditional moose material system has no support for nodal material property evaluation. The on-the-fly functor system does allow nodal evaluations, but it requires the user to specify a subdomain ID. This is to eliminate ambiguity about how to evaluate potentially discontinuous properties at subdomain interfaces

**Reply by hityyds** [reply]
Thanks for your detailed explanation!
